### PR TITLE
Fix multi worker

### PR DESCRIPTION
**Pull request by zhreshold**
## Description ##
- Fix race condition for CPUSharedStorageManager->Free
- Launch workers at iter init stage to avoid frequent relaunch

@piiswrong @sxjscience @yajiedesign 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by sxjscience** [comment]
@cjolivier01 @Jerryzcn

**Comment by Jerryzcn** [comment]
thank god

**Comment by Jerryzcn** [comment]
I will patch this and test on our speech dataset

File: python/mxnet/gluon/data/dataloader.py [hunk 1]
```diff
@@ -119,6 +119,71 @@ def worker_loop(dataset, key_queue, data_queue, batchify_fn):
...
+        for _ in range(self._num_workers):
+            worker = multiprocessing.Process(
+                target=worker_loop,
+                args=(self._dataset, self._key_queue, self._data_queue, self._batchify_fn))
+            worker.daemon = True
+            worker.start()
+            workers.append(worker)
+
+        for idx, batch in enumerate(self._batch_sampler):
+            self._key_queue.put((idx, batch))
```
Comment (sxjscience) [inline python/mxnet/gluon/data/dataloader.py]: May still need to revise the logic in later PRs to dynamically push more jobs into the key_queue.

**Comment by Jerryzcn** [comment]
can we add tests for the race condition?

**Comment by zhreshold** [comment]
@Jerryzcn The latest commit should fix #10042

**Pull request merged**